orttraining/orttraining/core/graph/gradient_schema_defs.cc

```diff
@@ -1339,36 +1373,44 @@ Return true if all elements are true and false otherwise.
       .SinceVersion(1)
       .SetSupportLevel(OpSchema::SupportType::EXPERIMENTAL)
       .SetDoc("Receive a tensor from the the specified source.")
-      .Input(0, "InputSignal", "Input control signal.", "TBool")
-      .Output(0, "OutputSignal", "Output control signal.", "TBool")
-      .Output(1, "Data", "The Received tensor.", "T")
-      .Attr("src",
-            "Abstractive memory ID of Data's source.",
-            AttributeProto::INT)
-      .Attr("dst",
-            "Abstractive memory ID of Data's destination.",
-            AttributeProto::INT)
+      .Input(0, "InputSignal", "Input control signal. It must be a scalar.", "TBool")
+      .Input(1, "Remote", "Remote src rank. It must be a scalar.", "TInt64")
+      .Output(0, "OutputSignal", "Output control signal. It must be a scalar.", "TBool")
+      .Output(1, "Data", "The Received tensors.", "V", OpSchema::Variadic, false)
       .Attr("tag", "The tag of the message carrying Data.",
             AttributeProto::INT)
-      .Attr("element_type", "Element type of the received tensor.",
-            AttributeProto::INT)
+      .Attr("element_types", "Element types of the received tensors.",
+            AttributeProto::INTS)
       .TypeConstraint(
-          "T",
-          {"tensor(float16)", "tensor(float)", "tensor(double)"},
-          "Constrain input types to float tensors.")
+          "TInt64",
+          {"tensor(int64)"},
+          "Constrain input type to 64-bit integer.")
       .TypeConstraint(
           "TBool",
           {"tensor(bool)"},
           "Constrain types to boolean tensors.")
+      .TypeConstraint("V", OpSchema::all_tensor_types(), "All Tensor types")
       .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
-        if (ctx.getNumInputs() != 1)
-          fail_shape_inference("Recv must have one inputs.");
-        if (ctx.getNumOutputs() != 2)
-          fail_shape_inference("Recv must have two output.");
+        if (ctx.getNumInputs() != 2) {
+          fail_shape_inference("Recv must have two inputs.");
+        } else {
+          auto& signal_input_shape = getInputShape(ctx, 0);
+          if (static_cast<int>(signal_input_shape.dim_size()) != 0) {
+            fail_shape_inference("InputSignal of Recv must be a scalar.");
+          }
+          auto& remote_input_shape = getInputShape(ctx, 1);
+          if (static_cast<int>(remote_input_shape.dim_size()) != 0) {
+            fail_shape_inference("Remote of Recv must be a scalar.");
+          }
+        }
+
+        if (ctx.getNumOutputs() < 2) {
```

Comment:
>getNumOutputs [](start = 16, length = 13)

Check number of data output == element_types.size() ?

Reply:
I did that in propagateRecvOutputTensorElemTypes for Recv.